src/dsp_tools/commands/xmlupload/xmlupload.py

```diff
@@ -343,129 +291,138 @@ def _extract_resources_from_xml(root: etree._Element, default_ontology: str) ->
+            f"In case of successful creation, call 'resume-xmlupload' with the flag "
+            f"'--skip-first-resource' to prevent duplication.\n"
+            f"If not, a normal 'resume-xmlupload' can be started."
+        )
+        logger.error(msg)
+        raise XmlUploadInterruptedError(msg) from None
+
+    try:
+        _tidy_up_resource_creation_idempotent(upload_state, iri, resource)
+        _interrupt_if_indicated(upload_state, creation_attempts_of_this_round)
+    except KeyboardInterrupt:
+        warnings.warn(DspToolsUserWarning("KeyboardInterrupt: Tidying up, then exit..."))
+        _tidy_up_resource_creation_idempotent(upload_state, iri, resource)
+        raise XmlUploadInterruptedError("KeyboardInterrupt during tidy up") from None
+
+
+def _interrupt_if_indicated(upload_state: UploadState, creation_attempts_of_this_round: int) -> None:
+    # if the interrupt_after value is not set, the upload will not be interrupted
+    interrupt_after = upload_state.config.interrupt_after or 999_999_999
+    if creation_attempts_of_this_round + 1 >= interrupt_after:
+        raise XmlUploadInterruptedError(
+            f"Interrupted: Maximum number of resources was reached ({upload_state.config.interrupt_after})"
+        )
+
+
+def _tidy_up_resource_creation_idempotent(
+    upload_state: UploadState,
+    iri: str | None,
     resource: XMLResource,
-    current_res: int,
-    total_res: int,
 ) -> None:
-    iri_resolver.update(resource.res_id, iri)
-    resource_designation = f"'{label}' (ID: '{resource.res_id}', IRI: '{iri}')"
-    print(f"{datetime.now()}: Created resource {current_res}/{total_res}: {resource_designation}")
-    logger.info(f"Created resource {current_res}/{total_res}: {resource_designation}")
+    previous_successful = len(upload_state.iri_resolver.lookup)
+    previous_failed = len(upload_state.failed_uploads)
+    upcoming = len(upload_state.pending_resources)
+    current_res = previous_successful + previous_failed + 1
+    total_res = previous_successful + previous_failed + upcoming
+    if iri:
+        # resource creation succeeded: update the iri_resolver
+        upload_state.iri_resolver.lookup[resource.res_id] = iri
+        msg = f"Created resource {current_res}/{total_res}: '{resource.label}' (ID: '{resource.res_id}', IRI: '{iri}')"
+        print(f"{datetime.now()}: {msg}")
+        logger.info(msg)
+    else:  # noqa: PLR5501
+        # resource creation failed gracefully: register it as failed
+        if resource.res_id not in upload_state.failed_uploads:
+            upload_state.failed_uploads.append(resource.res_id)
+
+    if resource in upload_state.pending_resources:
+        upload_state.pending_resources.remove(resource)
 
 
 def _create_resource(
     resource: XMLResource,
     bitstream_information: BitstreamInfo | None,
     resource_create_client: ResourceCreateClient,
-) -> tuple[str, str] | tuple[None, None]:
+) -> str | None:
```

Comment:
I simplified the return values of the ResourceCreateClient. It's useless that it returns the label, because we already know the label.